Nanoleaf-wpf/Api/Layouts/PanelLayout.cs

```diff
@@ -0,0 +1,173 @@
+        private static int _defaultTriangleSize = 148; //The default size in pixels of a triangle side as reported by the Nanoleaf API
+
+        private INanoleafClient _nanoleafClient;
+
+        private Layout _layout;
+
+        private GlobalOrientation _globalOrientation;
+
+        /// <summary>
+        /// Panels represented by the Nanoleaf coordinate system, unscaled and therefore unsuitable for display
+        /// </summary>
+        private List<DrawablePanel> _unscaledTriangles { get; set; }
+
+        public PanelLayout(Device device)
+        {
+            _nanoleafClient = NanoleafClient.GetClientForDevice(device);
+
+            _unscaledTriangles = new List<DrawablePanel>();
+
+            GetLayout();
+        }
+
+        public void GetLayout()
+        {
+            _layout = _nanoleafClient.LayoutEndpoint.GetLayout();
+            _globalOrientation = _nanoleafClient.LayoutEndpoint.GetGlobalOrientation();
+
+            if (_layout != null)
+            {
+                ConstructPanelsAsTriangles();
+            }
+        }
+
+        private void ConstructPanelsAsTriangles()
+        {
+            //Reverse every Y coordinate since Y = 0 means top for the canvas but for Nanoleaf it means bottom
+            foreach (var panelPosition in _layout.PanelPositions)
+            {
+                panelPosition.Y = panelPosition.Y * -1;
+            }
+
+            //Calculate the transform for the global orientation. All panels should rotate over the center of all panels
+            var minX = _layout.PanelPositions.Min(pp => pp.X);
+            var maxX = _layout.PanelPositions.Max(pp => pp.X);
+            var minY = _layout.PanelPositions.Min(pp => pp.Y);
+            var maxY = _layout.PanelPositions.Max(pp => pp.Y);
+            var globalRotationTransform = new RotateTransform(_globalOrientation.Value, Math.Abs(maxX - minX) / 2, Math.Abs(maxY - minY) / 2);
+
+            //Create the triangles
+            foreach (var panelPosition in _layout.PanelPositions)
+            {
+                CreateTriangle(panelPosition.X, panelPosition.Y, panelPosition.Orientation, panelPosition.PanelId, globalRotationTransform);
+            }
+
+            //Normalize the triangle positions such that the coordinates start at 0
+            double minTriangleX = _unscaledTriangles.SelectMany(p => p.Polygon.Points).Min(p => p.X);
+            double minTriangleY = _unscaledTriangles.SelectMany(p => p.Polygon.Points).Min(p => p.Y);
+
+            foreach (var triangle in _unscaledTriangles)
+            {
+                //Move MidPoint
+                triangle.MidPoint = new Point(triangle.MidPoint.X - minTriangleX, triangle.MidPoint.Y - minTriangleY);
+
+                //Move triangle
+                for (var i = 0; i < triangle.Polygon.Points.Count; i++)
+                {
+                    var x = triangle.Polygon.Points[i].X - minTriangleX;
```

Comment:
Is `-=` an option with it

Reply:
No, they are not modifiable at that point